control/Controller.php

```diff
@@ -76,6 +83,61 @@ public function init() {
 	}
 
 	/**
+	 * @var bool track if the callInit function is running
+	 */
+	private $callInitRunning = false;
+
+	/**
+	 * Mark call init as started
+	 */
+	protected function startCallInit() {
+		$this->callInitRunning = true;
+	}
+
+	/**
+	 * Mark call init as ended
+	 */
+	protected function endCallInit() {
+		$this->callInitRunning = false;
+	}
+
+	/**
+	 * @return bool Whether call init is running
+	 */
+	protected function isCallInitRunning() {
+		return $this->callInitRunning;
+	}
+
+	/**
+	 * A stand in function to protect the init function from failing to be called as well as providing before and
+	 * after hooks for the init function itself
+	 *
+	 * This should be called on all controllers before handling requests
+	 */
+	public function callInit() {
```

Comment:
I would have gone with `doInit` to be consistent with the other `doMethods`, but that's just a point of artistic difference. :D